src/webgl/p5.Framebuffer.js

```diff
@@ -0,0 +1,552 @@
+    this.target._renderer._curCamera = prevCam;
+  }
+
+  resizeCanvas(width, height) {
+    this.autoSized = false;
+    this.width = width;
+    this.height = height;
+    this._handleResize();
+  }
+
+  pixelDensity(density) {
+    if (density) {
+      this.autoSized = false;
+      this.density = density;
+      this._handleResize();
+    } else {
+      return this.density;
+    }
+  }
+
+  autoSized(autoSized) {
+    if (autoSized === undefined) {
+      return this.autoSized;
+    } else {
+      this.autoSized = autoSized;
+      this._handleResize();
+    }
+  }
+
+  _checkIfFormatsAvailable() {
+    const gl = this.gl;
+
+    if (
+      this.target.webglVersion === constants.WEBGL &&
+      !gl.getExtension('WEBGL_depth_texture')
+    ) {
+      throw new Error('Unable to create depth textures in this environment');
+    }
+
+    if (
+      this.target.webglVersion === constants.WEBGL &&
+      this.depthFormat === constants.FLOAT
+    ) {
+      console.warn(
+        'FLOAT depth format is unavailable in WebGL 1. ' +
+          'Defaulting to UNSIGNED_BYTE.'
+      );
+      this.depthFormat = constants.UNSIGNED_BYTE;
+    }
+
+    if (![
+      constants.UNSIGNED_BYTE,
+      constants.FLOAT,
+      constants.HALF_FLOAT
+    ].includes(this.format)) {
+      console.warn(
+        'Unknown Framebuffer format. ' +
+          'Please use UNSIGNED_BYTE, FLOAT, or HALF_FLOAT. ' +
+          'Defaulting to UNSIGNED_BYTE.'
+      );
+      this.format = constants.UNSIGNED_BYTE;
+    }
+    if (![
+      constants.UNSIGNED_BYTE,
+      constants.FLOAT
+    ].includes(this.depthFormat)) {
+      console.warn(
+        'Unknown Framebuffer depth format. ' +
+          'Please use UNSIGNED_BYTE or FLOAT. Defaulting to UNSIGNED_BYTE.'
+      );
+      this.depthFormat = constants.UNSIGNED_BYTE;
+    }
+
+    const support = checkWebGLCapabilities(this.target._renderer);
+    if (!support.float && this.format === constants.FLOAT) {
+      console.warn(
+        'This environment does not support FLOAT textures. ' +
+          'Falling back to UNSIGNED_BYTE.'
+      );
+      this.format = constants.UNSIGNED_BYTE;
+    }
+    if (!support.float && this.depthFormat === constants.FLOAT) {
+      console.warn(
+        'This environment does not support FLOAT depth textures. ' +
+          'Falling back to UNSIGNED_BYTE.'
+      );
+      this.depthFormat = constants.UNSIGNED_BYTE;
+    }
+    if (!support.halfFloat && this.format === constants.HALF_FLOAT) {
+      console.warn(
+        'This environment does not support HALF_FLOAT textures. ' +
+          'Falling back to UNSIGNED_BYTE.'
+      );
+      this.format = constants.UNSIGNED_BYTE;
+    }
+
+    if (
+      this.channels === constants.RGB &&
+      [constants.FLOAT, constants.HALF_FLOAT].includes(this.format)
+    ) {
+      console.warn(
+        'FLOAT and HALF_FLOAT formats do not work cross-platform with only ' +
+          'RGB channels. Falling back to RGBA.'
```

Comment:
Is this true? I thought you could have a single channel texture of any format?

Reply:
I think single texture channels could still be fine, it's specifically RGB + float that doesn't work everywhere -- Chrome is fine, but [Firefox only works with RGBA32F.](https://stackoverflow.com/questions/45379051/cant-render-on-floating-point-rgb-texture-on-firefox)